shell/install.ps1

```diff
@@ -0,0 +1,120 @@
+  "opam-2.2.0-rc1-x86_64-windows.exe" = "f2ec830a5706c45cb56a96713e296ef756c3f2904ca15f7c2ad0442916a9585fa1de8070208f2a6bb3a84dc74b677f946f5bc386c8ed1489da802b1d66a5e094";
+}
+
+Function DownloadAndCheck {
+  param (
+    [string]$OpamBinUrl,
+    [string]$OpamBinTmpLoc,
+    [string]$OpamBinName
+  )
+
+  if (-not $SHA512s.ContainsKey($OpamBinName)) {
+    throw "no sha"
+  }
+
+  [Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
+  (New-Object System.Net.WebClient).DownloadFile($OpamBinUrl, $OpamBinTmpLoc)
+  $Hash = (Get-FileHash -Path $OpamBinTmpLoc -Algorithm SHA512).Hash
+
+  if ($Hash -ne "$($SHA512s[$OpamBinName])") {
+    throw "Checksum mismatch, a problem occurred during download."
+  }
+}
+
+if (-not [System.Environment]::Is64BitOperatingSystem) {
+  throw "opam requires a 64-bit version of Windows"
+}
+
+if ($Dev.IsPresent) {
+  $Version = $DevVersion
+}
+
+$Tag = $Version.Replace("~", "-")
+$Arch = "x86_64"
+$OS = "windows"
+$OpamBinUrlBase = "https://github.com/ocaml/opam/releases/download/"
+$OpamBinName = "opam-$Tag-$Arch-$OS.exe"
+$OpamBinUrl = "$OpamBinUrlBase$Tag/$OpamBinName"
+
+$OpamBinTmpLoc = "$Env:TEMP\$OpamBinName"
+
+if (-not (Test-Path -Path $OpamBinTmpLoc -IsValid)) {
+  throw "Failed to determine a temporary path for downloading opam"
+}
+
+Write-Host "## Downloading opam $Version for Windows on x86_64"
+$esc = [char]27
+
+if ([string]::IsNullOrEmpty($OpamBinDir)) {
+  $OpamBinDir = Read-Host "## Where should it be installed? [$esc[1m$DefaultBinDir$esc[0m]"
+  if ([string]::IsNullOrEmpty($OpamBinDir)) {
+    $OpamBinDir = $DefaultBinDir
+  }
+}
+
+if (-not (Test-Path -Path $OpamBinDir -IsValid) -or $OpamBinDir -contains ';') {
+  throw "Destination given for installation is not a valid path"
+}
+
+# Check existing opam binaries
+$AllOpam = Get-Command -All -Name opam -CommandType Application -ErrorAction Ignore | ForEach-Object -MemberName Source
+foreach($OneOpam in $AllOpam) {
+  if ($OneOpam -ne "$OpamBinDir\opam.exe") {
+    throw "We detected another opam binary installed at '$OneOpam'. To ensure problems won't occur later, please uninstall it or remove it from the PATH"
+  }
+}
+
+DownloadAndCheck -OpamBinUrl $OpamBinUrl -OpamBinTmpLoc $OpamBinTmpLoc -OpamBinName $OpamBinName
+
+# Install the binary
+if (-not (Test-Path -Path $OpamBinDir -PathType Container)) {
+  [void](New-Item -Force -Path $OpamBinDir -Type Directory)
+}
+[void](Move-Item -Force -Path $OpamBinTmpLoc -Destination "$OpamBinDir\opam.exe")
+
+# Add the newly installed binary to PATH for this and future sessions
+$EnvTarget = If ($IsAdmin) {'MACHINE'} Else {'USER'}
+foreach ($loc in $EnvTarget, 'PROCESS') {
+  $PATH = [Environment]::GetEnvironmentVariable('PATH', $loc)
+  if (-not ($PATH -split ';' -contains $OpamBinDir)) {
+    [Environment]::SetEnvironmentVariable('PATH', "$OpamBinDir;$PATH", $loc)
+    if ($loc -eq 'PROCESS') {
```

Comment:
Oops, that should be `-ne`... the idea was the message is only displayed if the registry is actually updated... although maybe that's over-thinking it?

Reply:
The registry will be updated exactly once by the loop either machine or user.
It can still be after and use `$EnvTarget` rather than `$loc` in the `Write-Host`.
It's a trivial point.  Let's not spend more time on it. :-)

Reply:
It's not on a subsequent install though - if the user uses the script a second time (registry already updated), then the registry is updated zero times and at that point I was trying to suppress the message?

Reply:
Yes, very good point!